### DISCUSSION TITLE
Inserting into PK column with GENERATED ALWAYS

### DISCUSSION THREAD

**Discussion by whompratt** [diesel-rs/diesel#3915] [Q&A] [answered]
To clarify: this is using postgresql

I've search around for many combinations of what I think are relevant keywords for this query and am getting nowhere, so any help would be appreciated.

I have a table with 2 columns, `id` and `name`. The table's `up.sql` file is as follows:

```pgsql
# up.sql
CREATE TABLE IF NOT EXISTS orgs
(
    org_id   integer PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
    org_name text NOT NULL
); 
```

The issue I'm experiencing is around representing this table in a struct and then inserting new rows into it. The generated `schema.rs` file lists the data type of the `id` column as `Int4`, meaning the struct can't have `Option<i32>` as the data type for this column without warning of a "trait bound unsatisfied" error. When you try to insert a value it fails however, you get the error `"cannot insert a non-DEFAULT value into column \"org_id\""`.

What is the appropriate solution for dealing with columns such as this? What should the resultant schema.rs look like? and What should the struct look like for a table like this?

-----

extra context:

```rust
# schema.rs
diesel::table! {
    orgs (org_id) {
        org_id -> Int4,
        org_name -> Text,
    }
}
```

```rust
# table model
#[derive(Queryable, Insertable, Selectable, Serialize, Deserialize)]
#[diesel(table_name = orgs)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct Org {
    pub org_id: i32,
    pub org_name: String,
}
```

```rust
# insert function
diesel::insert_into(orgs)
    .values(&new_org)
    .execute(connection)
    .expect("Error creating new org.");
```

**Comment by whompratt** [comment, marked as answer]
To answer my own question:

I had 1 struct, `Org`, which defined a row object.
When inserting a new org, I was using a second struct, `NewOrg`, which held the required values for a new row.
However, just before the insert, I was using the first `Org` struct instead of the `NewOrg` struct to build the insert, which was just entirely human error.
To resolve, I had to switch over to using the `NewOrg` struct before inserting, and add "Insertable" to the derive macro, and add a `#[diesel(table_name = orgs)]` macro for it to be valid.
With these changes, the operations work as expected.

-----

New code:

```rust
# schema.rs
diesel::table! {
    orgs (org_id) {
        org_id -> Int4,
        org_name -> Text,
    }
}
```

```rust
# models.rs
#[derive(Queryable, Insertable, Selectable, Serialize, Deserialize)]
#[diesel(table_name = orgs)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct Org {
    pub org_id: i32,
    pub org_name: String,
}
```

```rust
#[derive(Serialize, Deserialize, Insertable)]
#[diesel(table_name = schema::orgs)]
pub struct NewOrg {
    org_name: String
}

pub fn create_org( ... {
    let connection = &mut establish_connection_pg();

    let new_org = NewOrg {
        org_name: org.org_name.to_string(),
    };

    diesel::insert_into(orgs)
        .values(&new_org)
        .execute(connection)
        .expect("Error creating new org.");
    Ok(Created::new("/").body(org))
}

**Reply by weiznich** [reply]
That's one solution that works. The other one is to use an explicit tuple insert:

```rust
diesel::insert_into(orgs)
     // that would be a tuple of values if you have several columns
    .values(schema::orgs::org_name.eq(&org.org_name))
    .execute(connection)?;
```

**Reply by whompratt** [reply]
> That's one solution that works. The other one is to use an explicit tuple insert:

That does seem like a more elegant solution.

-----

Currently I have functionally duplicate structs, one defining the object itself and one defining the object minus the generated columns to use for inserts.

Would something like the below be a reasonable implementation?
This way I'm using the existing model for an org, but instantiating with an `id` of 1 every time, then just... not inserting the `id`.

```Rust
#[derive(...)]
#[diesel(table_name = orgs)]
pub struct Org {
    pub org_id: i32,
    pub org_name: String,
}

let org: Org = Org {
    org_id: 1,
    org_name: input.org_name // wherever this comes from...
}

diesel::insert_into(orgs)
    .values(schema::orgs::org_name.eq(&org.org_name))
    .execute(connection)?;
```

**Reply by weiznich** [reply]
I generally advice people to use a seperate struct with `#[derive(Insertable)]` if that struct already exists because it either makes the API nicer or because you get some external data via e.g. json and you want to have that struct to put a `#[derive(Deserialize)]` onto. Otherwise I personally go with the tuple variant. I wouldn't use a solution that sets fields that don't exist. Prefer using a separate type for that.
